[RFC] Refactorings

utzig opened the pull request:
This adds some fixes and refactorings for multi-images. `boot_serial` support for multi-images is added with commit https://github.com/JuulLabs-OSS/mcuboot/commit/ceea2b933188edc2d7777e792b1436413324901d and related `newtmgr` (and `mcumgr` later) change is https://github.com/apache/mynewt-newtmgr/pull/143

utzig commented on boot/bootutil/src/encrypted.c:
```diff
@@ -242,23 +241,15 @@ boot_enc_load(struct enc_key_data *enc_state, int image_index,
         rc = flash_area_read(fap, off, &tlv, sizeof tlv);
         if (rc) {
-            return rc;
+            return -1;
```
I felt like it could be returning a positive value and not generating an error, but I think I will revert back to returning `rc`.

nvlsianpu replied:
I'm fine with -1.  Maybe worth to add note on error-codes to the porting manual.

utzig replied:
Reverted back to returning `rc` (should be fine).

utzig commented on boot/bootutil/src/loader.c:
```diff
@@ -242,6 +242,36 @@ boot_status_source(struct boot_loader_state *state)
 }
 
+/*
+ * Locate the TLVs in an image.
```
Ok, gonna do, but `hdr` and `fap` are used everywhere so there's nothing special about them here! :-P

nvlsianpu replied:
I asked about this as this is new code, no need for document this in not touched areas.

utzig replied:
Done.

d3zd3z commented:
Looks really good. The only suggestion I'd have would be to swap the "Fix dependency check infinite loop" and "Uncomment hold and add correct upgrades", since as they are, if a bisect hits the "Uncomment..." change, the test will lock up.

utzig commented:
Btw, had to update 020b707d3369275669238ff08ee7f650788b1058 to leave `boot_swap_type` there because `mcumgr` and `newtmgr` need it, while they have not been updated for multi-image compatibility ("We don't break the userspace"!). The bootloader itself still only uses `boot_swap_type_multi(<image-number>)`.

utzig commented:
@nvlsianpu @d3zd3z I removed the "WIP" here; the issues were addressed, I think it will be more productive to merge this and focus on thoroughly reviewing the dependencies and encrypted multi-images afterwards. I removed the WIP in the commit that fixes the dependencies on the simulator (it looks ok, I just rewrote the test to follow the previous logic). I added a few commits, with fixes to swap (as reported on #427) and related documentation, plus a few formatting issues...

nvlsianpu commented on boot/bootutil/src/loader.c:
```diff
@@ -1831,10 +1831,12 @@ boot_verify_slot_dependency(struct boot_loader_state *state,
      * the dependency and get it's version. */
-    dep_slot = (state->swap_type[dep->image_id] != BOOT_SWAP_TYPE_NONE) ?
+    swap_type = state->swap_type[dep->image_id];
+    dep_slot = (swap_type != BOOT_SWAP_TYPE_NONE && swap_type != BOOT_SWAP_TYPE_FAIL) ?
```
I would add comment on reason for include BOOT_SWAP_TYPE_FAIL: "When only upgrading one slot, the swap type for the non upgraded slot is
BOOT_SWAP_TYPE_FAIL instead of NONE."
